internal/controller/nemoservice_controller.go

```diff
@@ -0,0 +1,586 @@
+				logger.Error(err, "Failed to reconcile NeMo service", "service", serviceType, "name", serviceSpec.Name)
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// Update status of NeMoService
+	if err := r.updateStatus(ctx, nemoService, enabledServices); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Cleanup disabled services
+	if err := r.cleanupDisabledNemoMicroservices(ctx, nemoService, enabledServices); err != nil {
+		logger.Error(err, "Failed to cleanup disabled NeMo microservices")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Generic function to set spec for all NeMo service types
+func SetObjectSpec[T any](obj client.Object, service NemoServiceSpec[T]) error {
+	// Ensure obj implements runtime.Object
+	if _, ok := obj.(runtime.Object); !ok {
+		return fmt.Errorf("obj does not implement runtime.Object")
+	}
+
+	// Extract the spec from the generic service type
+	spec := service.Spec
+
+	// Handle all NeMo service types
+	switch t := obj.(type) {
+	case *appsv1alpha1.NemoCustomizer:
+		serviceSpec, ok := any(spec).(appsv1alpha1.NemoServiceCustomizerSpec)
+		if !ok {
+			return fmt.Errorf("service spec is not of type NemoServiceCustomizerSpec")
+		}
+		t.Spec = *serviceSpec.Spec
+
+	case *appsv1alpha1.NemoDatastore:
+		serviceSpec, ok := any(spec).(appsv1alpha1.NemoServiceDatastoreSpec)
+		if !ok {
+			return fmt.Errorf("service spec is not of type NemoServiceDatastoreSpec")
+		}
+		t.Spec = *serviceSpec.Spec
+
+	case *appsv1alpha1.NemoEvaluator:
+		serviceSpec, ok := any(spec).(appsv1alpha1.NemoServiceEvaluatorSpec)
+		if !ok {
+			return fmt.Errorf("service spec is not of type NemoServiceEvaluatorSpec")
+		}
+		t.Spec = *serviceSpec.Spec
+
+	case *appsv1alpha1.NemoGuardrail:
+		serviceSpec, ok := any(spec).(appsv1alpha1.NemoServiceGuardrailsSpec)
+		if !ok {
+			return fmt.Errorf("service spec is not of type NemoServiceGuardrailSpec")
+		}
+		t.Spec = *serviceSpec.Spec
+
+	case *appsv1alpha1.NemoEntitystore:
+		serviceSpec, ok := any(spec).(appsv1alpha1.NemoServiceEntitystoreSpec)
+		if !ok {
+			return fmt.Errorf("service spec is not of type NemoServiceEntitystoreSpec")
+		}
+		t.Spec = *serviceSpec.Spec
+
+	default:
+		return fmt.Errorf("unsupported object type: %T", obj)
+	}
+	return nil
+}
+
+// Generic function for reconciling any Nemo service
+func reconcileGenericService[T any](
+	ctx context.Context,
+	r *NemoServiceReconciler,
+	nemoService *appsv1alpha1.NemoService,
+	service NemoServiceSpec[T],
+	obj client.Object,
+) error {
+	logger := log.FromContext(ctx)
+
+	if service.Enabled == nil || !*service.Enabled {
+		return nil
+	}
+
+	// Set metadata
+	obj.SetName(service.Name)
+	obj.SetNamespace(nemoService.Namespace)
+
+	// Set service spec
+	err := SetObjectSpec(obj, service)
+	if err != nil {
+		logger.Error(err, "Failed to set object spec")
+		return err
+	}
+
+	// Set controller reference
+	if err := controllerutil.SetControllerReference(nemoService, obj, r.Scheme); err != nil {
```

Comment:
I believe we should wait for all child CRs to be deleted before the parent CR is deleted here.
```suggestion
	if err := controllerutil.SetControllerReference(nemoService, obj, r.Scheme, controllerutil.WithBlockOwnerDeletion(true)); err != nil {
```